packages/connectors/connector-sendgrid-email/src/index.ts

```diff
@@ -16,42 +19,84 @@ import {
 } from '@logto/connector-kit';
 
 import { defaultMetadata, endpoint } from './constant.js';
-import { sendGridMailConfigGuard } from './types.js';
-import type { EmailData, Personalization, Content, PublicParameters } from './types.js';
+import { ContextType, sendGridMailConfigGuard } from './types.js';
+import type { PublicParameters, SendGridMailConfig } from './types.js';
+
+const buildDefaultTemplateParameters = (
+  to: string,
+  config: SendGridMailConfig,
+  template: SendGridMailConfig['templates'][0],
+  payload: SendMessagePayload
+): PublicParameters => {
+  return {
+    personalizations: [{ to: [{ email: to }] }],
+    from: {
+      email: config.fromEmail,
+      ...conditional(config.fromName && { name: config.fromName }),
+    },
+    subject: replaceSendMessageHandlebars(template.subject, payload),
+    content: [
+      {
+        type: template.type,
+        value: replaceSendMessageHandlebars(template.content, payload),
+      },
+    ],
+  };
+};
+
+const buildCustomTemplateParameters = (
+  to: string,
+  config: SendGridMailConfig,
+  { subject, content, replyTo, sendFrom, contentType = 'text/html' }: EmailTemplateDetails,
+  payload: SendMessagePayload
+): PublicParameters => {
+  return {
+    personalizations: [
+      {
+        to: [
+          {
+            email: to,
+            // If replyTo is provided, we will replace the handlebars with the payload
+            ...conditional(replyTo && { name: replaceSendMessageHandlebars(replyTo, payload) }),
+          },
+        ],
+      },
+    ],
+    from: {
+      email: config.fromEmail,
+      // If sendFrom is provided, we will replace the handlebars with the payload
+      ...conditional(sendFrom && { name: replaceSendMessageHandlebars(sendFrom, payload) }),
+    },
+    subject: replaceSendMessageHandlebars(subject, payload),
+    content: [
+      {
+        type: contentType === 'text/html' ? ContextType.Html : ContextType.Text,
+        value: replaceSendMessageHandlebars(content, payload),
+      },
+    ],
+  };
+};
 
 const sendMessage =
-  (getConfig: GetConnectorConfig): SendMessageFunction =>
+  (
+    getConfig: GetConnectorConfig,
+    getI18nEmailTemplate?: GetI18nEmailTemplate
```

Comment:
How about `getCustomEmailTemplate`, because the connector does not know that is "i18n"

Reply:
This function contains a locale input parameter that indicates the i18n context.